### PR TITLE
Image decoder

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>
Co-authored-by: Joaquin Anton <janton@nvidia.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
It adds a high-level decoder that aggregates all decoders.

The ImageDecoder implements the ImageDecoderInstance interface, just as specific decoders do. It also implements ImageParser, because it can also parse arbitrary images.

ImageDecoder aggreates several decoders, filtered by a predicate. These decoders are then selected based on compatibility with image format and priority,

Operation:
For each decoder, there's a worker thread which picks up a sub-batch of work and calls `ScheduleDecode` on the associated decoder, then it waits results.
If a decoder has a fallback, then the worker waits for partial results. If there are failures, the failing samples are sent to the fallback worker. The successes are propagated to the top-most result promise.
If a decoder doesn't have a fallback, the worker waits for all results and propagates them to the top-most results promise, when ready.

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2984

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: dali/imgcodec/decoders/libjpeg_turbo.cc [hunk 1]
```diff
@@ -67,19 +67,13 @@ DecodeResult LibJpegTurboDecoderInstance::DecodeImplTask(int thread_idx,
...
-    if ((res.success = decoded_image != nullptr)) {
-      // JPEG images are always 8-bit, in HWC format
-      SampleView<CPUBackend> in(decoded_image.get(), target_shape, DALI_UINT8);
-      TensorLayout layout = "HWC";
-      Convert(out, layout, out_type, in, layout, flags.color_space, {}, {});
-    }
-  } catch (...) {
-    res.exception = std::current_exception();
-    res.success = false;
+  auto decoded_image = jpeg::Uncompress(encoded_data, data_size, flags);
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/libjpeg_turbo.cc:70]: no need to try/catch anymore

File: dali/imgcodec/decoders/libtiff/tiff_libtiff.cc [hunk 2]
```diff
@@ -51,8 +51,10 @@ class DecoderHelper {
 
   static toff_t seek(thandle_t handle, toff_t offset, int whence) {
     DecoderHelper *helper = reinterpret_cast<DecoderHelper *>(handle);
-    helper->stream_->SeekRead(offset, whence);
-    return helper->stream_->TellRead();
+    try {
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/libtiff/tiff_libtiff.cc:54]: this will go away (part of another PR)

File: dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc [hunk 3]
```diff
@@ -238,23 +238,18 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
...
     (ctx.bpp != 8 && ctx.bpp != 16);
 
-  try {
-    auto decode_out = out;
-    if (is_processing_needed) {
-      int64_t type_size = dali::TypeTable::GetTypeInfo(ctx.pixel_type).size();
-      res.intermediate_buffer.resize(volume(ctx.shape) * type_size);
-      decode_out = {res.intermediate_buffer.data(), ctx.shape, ctx.pixel_type};
-    }
+  auto decode_out = out;
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc:241]: no need to try/catch

**Comment by mzient** [comment]
!build

File: include/dali/imgcodec/image_decoder.h [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                      DecodeParams opts,
+                      const ROI &roi = {}) override;
+
+  DecodeResult Decode(DecodeContext ctx,
+                      SampleView<GPUBackend> out,
+                      ImageSource *in,
+                      DecodeParams opts,
+                      const ROI &roi = {}) override;
+
+  FutureDecodeResults ScheduleDecode(DecodeContext ctx,
```
Comment (prak-nv) [inline include/dali/imgcodec/image_decoder.h:119]: AsyncDecode would be more cannonical name.

File: dali/imgcodec/image_decoder.cc [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+  for (auto &[f, w] : dist) {
+    auto &worker = workers_[filtered_.find(f)->second];
+    worker->add_work(std::move(w));
+  }
+}
+
+void ImageDecoder::InitWorkers(bool lazy_init) {
+  const ImageFormat *prev_format = nullptr;
+  DecoderWorker *prev_worker = nullptr;
+  for (auto [format, factory] : filtered_) {
```
Comment (prak-nv) [inline dali/imgcodec/image_decoder.cc:282]: Why do we have DecoderWorker active objects per format; rather than simply use thread_pool of configurable size and schedule work to be done there?

**Review by prak-nv — APPROVED** [review]
LGTM, just don't get why do we have to do it this way.

File: dali/imgcodec/image_decoder.cc [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+      }
+    } else {
+      // no fallback - the results are final
+      future.wait_all();
+      auto r = future.get_all();
+      // we may need to copy to the gpu
+      if (!decode_to_gpu && !work->gpu_outputs.empty()) {
+        for (int i = 0; i < r.size(); i++) {
+          if (r[i].success) {  // TODO(michalz): bulk copy
+            copy(work->gpu_outputs[i], work->cpu_outputs[i], work->ctx.stream);
```
Comment (szkarpinski) [inline dali/imgcodec/image_decoder.cc:476]: This copy will happen only if a fallback was used. As a result, when we successfully decode whole back in the first attempt using CPU-only decoder (libtiff, libjpeg-turbo), no copy to GPU will happen.

We should probably have a copy in the other branch too.

File: dali/imgcodec/image_decoder.cc (hunk 5 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:282]: There are multiple reasons:
1. The workers schedule work for one another. The decoder fallback system is a DAG, so there's no risk of deadlock. While the number of decoders is fixed, the number of work items in the queues is not guaranteed not to exceed the number of decoders (pessimistically it can reach the batch size). In that case, it's possible that a cycle would form.
2. The work is always the same - only the data changes, but it's always the same function. only the data change. The data objects are kept in a pool to avoid extra allocations.
3. ScheduleDecode in `decoder_` is not thread-safe. Here, there's one dedicated thread that calls it.

I'm not saying that this work scheme could not be abstracted - it could, but that'd come with additional development time cost which we cannot afford at this point.

File: dali/imgcodec/image_decoder.cc (hunk 6 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:484]: Fixed.

**Comment by mzient** [comment]
!build

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 7]
```diff
@@ -76,6 +76,43 @@ static void AssertEqualSatNorm(const Tensor<CPUBackend> &img,
 }
 
 
+
+template <typename OutputType, typename RefType>
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/decoder_test_helper.h:80]: I'd appreciate some doxygen here. It's not clear what "similar" means only by the name

File: dali/pipeline/util/thread_pool.cc [hunk 8]
```diff
@@ -60,20 +60,19 @@ ThreadPool::~ThreadPool() {
 }
 
 void ThreadPool::AddWork(Work work, int64_t priority, bool start_immediately) {
-  bool started_before = false;
+  bool should_start_all = false;
```
Comment (jantonguirao) [inline dali/pipeline/util/thread_pool.cc:63]: I think we should make this a separate PR, for better traceability.

File: include/dali/imgcodec/image_decoder.h [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+   */
+  ImageDecoder(int device_id,
+               bool lazy_init,
+               const std::map<std::string, any> &params,
+               std::function<bool(ImageDecoderFactory *)> decoder_filter);
+
+  ~ImageDecoder();
+
+  bool CanParse(ImageSource *encoded) const override {
+    auto *f = FormatRegistry().GetImageFormat(encoded);
```
Comment (jantonguirao) [inline include/dali/imgcodec/image_decoder.h:70]: why not move all implementations to the cc file?

File: include/dali/imgcodec/image_decoder.h [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    void alloc_temp_cpu_outputs(ImageDecoder &owner);
+
+    DecodeContext ctx;
+    // The original promise
+    DecodeResultsPromise results;
+    // The indices in the original request
+    BatchVector<int> indices;
+    BatchVector<ImageSource *> sources;
+    BatchVector<SampleView<CPUBackend>> cpu_outputs;
+    BatchVector<SampleView<GPUBackend>> gpu_outputs;
```
Comment (jantonguirao) [inline include/dali/imgcodec/image_decoder.h:247]: how about having a single vector a struct with index, source, cpu_out, gpu_out...

File: include/dali/imgcodec/image_decoder.h [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+  int device_id_ = CPU_ONLY_DEVICE_ID;
+  std::map<std::string, any> params_;
+  std::map<const ImageDecoderFactory*, std::unique_ptr<DecoderWorker>> workers_;
+  std::multimap<const ImageFormat*, ImageDecoderFactory*> filtered_;
+  std::vector<std::shared_ptr<void>> temp_buffers_;
+  struct TempImageSource {
+    // TODO(michalz): store auxiliary image sources here
+  };
+  std::map<InputKind, std::vector<TempImageSource>> tmp_sources_;
```
Comment (jantonguirao) [inline include/dali/imgcodec/image_decoder.h:360]: Why do we need this?

File: dali/imgcodec/image_decoder.cc [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+    for (size_t i = 0; i < mask.size(); i++) {
+      if (!mask[i])
+        work->results.set(work->indices[i], DecodeResult::Failure(nullptr));
+    }
+  }
+
+  if (!work->sources.empty()) {
+    bool decode_to_gpu = produces_gpu_output_;
+
+    if (!decode_to_gpu && work->cpu_outputs.empty()) {
```
Comment (jantonguirao) [inline dali/imgcodec/image_decoder.cc:434]: are we sure that pre-existing `cpu_outputs` have the right shape?

File: dali/imgcodec/image_decoder.cc [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+  for (auto &o : out)
+    work->cpu_outputs.push_back(o);
+  for (auto *src : in)
+    work->sources.push_back(src);
+  for (int i = 0; i < N; i++)
+    work->indices.push_back(i);
+
+  DistributeWork(std::move(work));
+
+  temp_buffers_.clear();
```
Comment (jantonguirao) [inline dali/imgcodec/image_decoder.cc:220]: I am struggling to understand why you need to clear again here. Can you explain?

File: include/dali/imgcodec/image_decoder.h [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+               const std::map<std::string, any> &params = {})
+  : ImageDecoder(
+      device_id,
+      lazy_init,
+      params,
+      device_id == CPU_ONLY_DEVICE_ID ? is_cpu_decoder
+                                      : [](ImageDecoderFactory *) { return true; }) {}
+
+  /**
+   * @brief Construct a new ImageDecode, allowing the user to filter the decoders
```
Comment (jantonguirao) [inline include/dali/imgcodec/image_decoder.h:51]:
```suggestion
   * @brief Construct a new ImageDecoder, allowing the user to filter the decoders
```

File: dali/imgcodec/image_decoder.cc (hunk 5 above)
Comment (jantonguirao) [reply dali/imgcodec/image_decoder.cc:288]: What's the point of keeping temporary buffers alive here instead of relying on the memory resource?

File: dali/imgcodec/decoders/decoder_test_helper.h (hunk 7 above)
Comment (mzient) [reply dali/imgcodec/decoders/decoder_test_helper.h:80]: I can add something.

File: dali/imgcodec/image_decoder.cc (hunk 5 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:288]: @jantonguirao I think this comment is misplaced.

File: dali/imgcodec/decoders/decoder_test_helper.h (hunk 7 above)
Comment (mzient) [reply dali/imgcodec/decoders/decoder_test_helper.h:86]: Done.

File: dali/pipeline/util/thread_pool.cc (hunk 8 above)
Comment (mzient) [reply dali/pipeline/util/thread_pool.cc:63]: Yeah, I guess.

File: include/dali/imgcodec/image_decoder.h (hunk 9 above)
Comment (mzient) [reply include/dali/imgcodec/image_decoder.h:69]: Done.

File: include/dali/imgcodec/image_decoder.h (hunk 14 above)
Comment (mzient) [reply include/dali/imgcodec/image_decoder.h]: Fixed.

File: dali/imgcodec/image_decoder.cc (hunk 13 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc]: Moved to ScheduledWork.

File: include/dali/imgcodec/image_decoder.h [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+   *
+   * A DecoderWorker waits for incoming ScheduledWork objects and processes them by running
+   * `decoder_->ScheduleDecode` and waiting for partial results, scheduling the failed
+   * samples to a fallback decoder, if present.
+   *
+   * When a sample is successfully decoded, it is marked as succeess in the parent
+   * DecodeResultsPromise. If it fails, it goes to fallback and only if all fallbacks fail, it is
+   * marked in the DecodeResultsPromise as a failure.
+   */
+  class DecoderWorker {
```
[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/imgcodec/image_decoder.h]: Moved.

File: include/dali/imgcodec/image_decoder.h [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  /**
+   * @brief A worker that processes sub-batches of work to be processed by a particular decoder.
+   *
+   * A DecoderWorker waits for incoming ScheduledWork objects and processes them by running
+   * `decoder_->ScheduleDecode` and waiting for partial results, scheduling the failed
+   * samples to a fallback decoder, if present.
+   *
+   * When a sample is successfully decoded, it is marked as succeess in the parent
```
[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/imgcodec/image_decoder.h]: Fixed.

File: dali/imgcodec/image_decoder.cc (hunk 5 above)
Comment (jantonguirao) [reply dali/imgcodec/image_decoder.cc:630]: It was a general question, following your explanation above

File: include/dali/imgcodec/image_decoder.h (hunk 10 above)
Comment (mzient) [reply include/dali/imgcodec/image_decoder.h]: Nope, becasue we need spans of the separate entities to pass them to SchedulDecode.

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/image_decoder.cc [hunk 17]
```diff
@@ -0,0 +1,772 @@
...
+      auto ptr = std::move(free_work_items_);
+      free_work_items_ = std::move(ptr->next);
+      ptr->results = std::move(results);
+      ptr->ctx = std::move(ctx);
+      ptr->params = std::move(params);
+      return ptr;
+    }
+  }
+
+  return std::make_unique<ScheduledWork>(ctx, results);
```
Comment (szkarpinski) [inline dali/imgcodec/image_decoder.cc:738]: As far as I understand we should set `.params` also for fresh work items

File: dali/imgcodec/image_decoder.cc (hunk 12 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc]: Of course they don't because there aren't any.

File: dali/imgcodec/image_decoder.cc (hunk 17 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc]: Fixed.

File: dali/imgcodec/image_decoder.cc [hunk 18]
```diff
@@ -0,0 +1,772 @@
...
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  return ScheduleDecode(ctx, make_span(&out, 1), make_span(&in, 1), opts, make_span(&roi, 1));
+}
+
+FutureDecodeResults ImageDecoder::ScheduleDecode(DecodeContext ctx,
+                                                span<SampleView<CPUBackend>> out,
+                                                cspan<ImageSource *> in,
+                                                DecodeParams opts,
+                                                cspan<ROI> rois) {
```
Comment (szkarpinski) [inline dali/imgcodec/image_decoder.cc:546]: We ignore the `rois` parameter here (except for asserting its correctness :P), probably we'd like to copy it to `work->rois`

Comment (mzient) [reply dali/imgcodec/image_decoder.cc]: Fixed in ScheduledWork::init

**Comment by mzient** [comment]
!build

File: dali/imgcodec/image_decoder.cc [hunk 19]
```diff
@@ -0,0 +1,782 @@
...
+    dst.reserve(dst.size() + src.size());
+    for (auto &x : src)
+      dst.emplace_back(std::move(x));
+  };
+
+  move_append(target.cpu_outputs, source->cpu_outputs);
+  move_append(target.gpu_outputs, source->gpu_outputs);
+  move_append(target.sources, source->sources);
+  move_append(target.indices, source->indices);
+  move_append(target.rois, source->rois);
```
Comment (mzient) [inline dali/imgcodec/image_decoder.cc:777]:
```suggestion
  move_append(target.rois, source->rois);
  move_append(target.temp_buffers, source->temp_buffers);
```

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/imgcodec/decoders/decoder_parallel_impl.h [hunk 20]
```diff
@@ -45,19 +45,9 @@ class DLL_PUBLIC BatchParallelDecoderImpl : public ImageDecoderImpl {
...
-
-    // writing in parallel to adjacent entries in vector<bool> is not safe
-    std::vector<int> tmp(in.size());
     ROI no_roi;
-    for (int i = 0; i < in.size(); i++) {
-      ctx.tp->AddWork([&, ctx, i](int) {
-        tmp[i] = CanDecode(ctx, in[i], opts, rois.empty() ? no_roi : rois[i]);
-      });
-    }
-    ctx.tp->RunAll();
```
Comment (mzient) [inline dali/imgcodec/decoders/decoder_parallel_impl.h:57]: This would hang, so I serialized the loop.

File: include/dali/imgcodec/image_decoder.h [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+
+  static void move_to_fallback(ScheduledWork *fallback,
+                               ScheduledWork &work,
+                               const vector<bool> &keep);
+
+  static void filter(ScheduledWork &work, const vector<bool> &keep) {
+    move_to_fallback(nullptr, work, keep);
+  }
+
+  void alloc_temp_cpu_outputs(ScheduledWork &work);
```
Comment (mzient) [inline include/dali/imgcodec/image_decoder.h:211]: to remove

File: dali/imgcodec/image_decoder.cc [hunk 22]
```diff
@@ -0,0 +1,791 @@
...
+    size_t size = volume(gpu_out.shape()) * TypeTable::GetTypeInfo(gpu_out.type()).size();
+    constexpr int kTempBufferAlignment = 256;
+
+    temp_buffers.emplace_back();
+    auto &buf_ptr = temp_buffers.back();
+    buf_ptr = mm::alloc_raw_async_unique<char, mm::memory_kind::pinned>(
+      size, mm::host_sync, ctx.stream, kTempBufferAlignment);
+
+    SampleView<CPUBackend> &cpu_out = cpu_outputs[i];
+    cpu_out = SampleView<CPUBackend>(buf_ptr.get(), gpu_out.shape(), gpu_out.type());
```
Comment (mzient) [inline dali/imgcodec/image_decoder.cc:159]: layout?

File: dali/imgcodec/image_decoder.cc [hunk 23]
```diff
@@ -0,0 +1,791 @@
...
+                                 work->params,
+                                 make_span(work->rois));
+
+    if (fallback_) {
+      for (;;) {
+        auto indices = future.wait_new();
+        if (indices.empty())
+          break;
+
+        auto r = future.get_all_ref();
```
Comment (mzient) [inline dali/imgcodec/image_decoder.cc:359]: This will wait for all!

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/image_decoder.cc [hunk 24]
```diff
@@ -0,0 +1,797 @@
...
+    }
+    cv_.notify_all();
+    worker_.join();
+    worker_ = {};
+  }
+}
+
+void ImageDecoder::DecoderWorker::run() {
+  DeviceGuard dg(owner_->device_id_);
+  while (!stop_requested_) {
```
Comment (banasraf) [inline dali/imgcodec/image_decoder.cc:274]: Can we safely read the value without locking the mutex?

File: dali/imgcodec/image_decoder.cc [hunk 25]
```diff
@@ -0,0 +1,797 @@
...
+  return Decode(ctx, make_span(samples), in, opts, rois);
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Work item management
+////////////////////////////////////////////////////////////////////////////
+
+std::unique_ptr<ImageDecoder::ScheduledWork> ImageDecoder::new_work(
+    DecodeContext ctx, DecodeResultsPromise results, DecodeParams params) {
+  if (free_work_items_) {
```
Comment (banasraf) [inline dali/imgcodec/image_decoder.cc:744]: Again, is it safe to read this value without the lock? It's not atomic, right?

File: dali/imgcodec/image_decoder.cc (hunk 24 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:274]: It's rechecked inside the lock, so nothing wrong should happen - hopefully we won't need to reacquire the lock if stop was requested while processing (but that's rare enough not to bother with).

File: dali/imgcodec/image_decoder.cc [hunk 26]
```diff
@@ -0,0 +1,797 @@
...
+    assert((work->cpu_outputs.empty() && work->gpu_outputs.size() == work->sources.size()) ||
+           (work->gpu_outputs.empty() && work->cpu_outputs.size() == work->sources.size()));
+    assert(work->rois.empty() || work->rois.size() == work->sources.size());
+    assert(work->temp_buffers.empty() || work->temp_buffers.size() == work->cpu_outputs.size());
+    if (work_) {
+      owner_->combine_work(*work_, std::move(work));
+      // no need to notify - a work item was already there, so it will be picked up regardless
+    } else {
+      work_ = std::move(work);
+      cv_.notify_one();
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:301]: True. It was more out of habit that things like "request stop" should make triple sure that no thread will be left hanging. It's not like notify_all is a problem - I could even use it here.

File: dali/imgcodec/image_decoder.cc (hunk 25 above)
Comment (mzient) [reply dali/imgcodec/image_decoder.cc:744]: As you can see, it's rechecked inside the lock. Such double checking is an established pattern.

**Review by banasraf — APPROVED** [review]
